primitives/src/settlement.rs

```diff
@@ -0,0 +1,437 @@
+    fn from(leg: Leg) -> Self {
+        LegV2 {
+            from: leg.from,
+            to: leg.to,
+            asset: LegAsset::Fungible {
+                ticker: leg.asset,
+                amount: leg.amount,
+            },
+        }
+    }
+}
+
+/// Details about a venue.
+#[derive(Encode, Decode, TypeInfo)]
+#[derive(Clone, Default, PartialEq, Eq, Debug, PartialOrd, Ord)]
+pub struct Venue {
+    /// Identity of the venue's creator
+    pub creator: IdentityId,
+    /// Specifies type of the venue (Only needed for the UI)
+    pub venue_type: VenueType,
+}
+
+/// Details about an offchain transaction receipt
+#[derive(Encode, Decode, Clone, PartialEq, Eq, Debug, PartialOrd, Ord)]
+pub struct Receipt<Balance> {
+    /// Unique receipt number set by the signer for their receipts
+    pub receipt_uid: u64,
+    /// Identity of the sender
+    pub from: PortfolioId,
+    /// Identity of the receiver
+    pub to: PortfolioId,
+    /// Ticker of the asset being transferred
+    pub asset: Ticker,
+    /// Amount being transferred
+    pub amount: Balance,
+}
+
+/// A wrapper for VenueDetails
+#[derive(Encode, Decode, TypeInfo, VecU8StrongTyped)]
+#[derive(Clone, Debug, Default, PartialEq, Eq, PartialOrd, Ord)]
+pub struct ReceiptMetadata(Vec<u8>);
+
+/// Details about an offchain transaction receipt that a user must input
+#[derive(Encode, Decode, TypeInfo, Clone, PartialEq, Eq, Debug, PartialOrd, Ord)]
+pub struct ReceiptDetails<AccountId, OffChainSignature> {
+    /// Unique receipt number set by the signer for their receipts
+    pub receipt_uid: u64,
+    /// Target leg id
+    pub leg_id: LegId,
+    /// Signer for this receipt
+    pub signer: AccountId,
+    /// signature confirming the receipt details
+    pub signature: OffChainSignature,
+    /// Generic text that can be used to attach messages to receipts
+    pub metadata: ReceiptMetadata,
+}
+
+/// Stores the number of fungible and non fungible transfers in a set of legs.
+#[derive(Default)]
+pub struct TransferData {
+    fungible: u32,
+    non_fungible: u32,
+}
+
+impl TransferData {
+    /// Creates an instance of `TransfersData`.
+    pub fn new(fungible: u32, non_fungible: u32) -> Self {
+        TransferData {
+            fungible,
+            non_fungible,
+        }
+    }
+
+    /// Returns the number of fungible transfers.
+    pub fn fungible(&self) -> u32 {
+        self.fungible
+    }
+
+    /// Returns the number of non fungible transfers.
+    pub fn non_fungible(&self) -> u32 {
+        self.non_fungible
+    }
+
+    /// Adds one to the number of fungible transfers.
+    pub fn add_fungible(&mut self) {
+        self.fungible += 1;
+    }
+
+    /// Adds `nfts.len()` to the number of non fungible transfers.
+    pub fn add_non_fungible(&mut self, nfts: &NFTs) {
+        self.non_fungible += nfts.len() as u32;
+    }
+
+    /// Gets the `TransferData` from a slice of `LegV2`.
+    pub fn from_legs(legs_v2: &[LegV2]) -> Result<TransferData, String> {
```

Comment:
This is the only block of code that actually changed besides the import paths. This was moved here to avoid code duplication, since it's used for the `WeightInfo` trait and in a single place in an extrinsic call. The code was changed to return a `Result`. Unlike `get_transfer_data`, which is only called after the checks for maximum size have been made, this one is not.